src/lib.rs

```diff
@@ -0,0 +1,238 @@
+// Copyright (c) 2019 King's College London created by the Software Development Team
+// <http://soft-dev.org/>
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0>, or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, or the UPL-1.0 license <http://opensource.org/licenses/UPL>
+// at your option. This file may not be copied, modified, or distributed except according to those
+// terms.
+
+#![feature(alloc_layout_extra)]
+
+use std::alloc::{alloc, Layout};
+use std::mem::{size_of, transmute};
+use std::time::Instant;
+
+const REPS: usize = 100;
+const VEC_SIZE: usize = 10000000;
+const VAL_NOREAD: usize = 0xdeadbeef;
+const VAL_WITHREAD: usize = 0xFEEDC0DE;
+
+#[inline(never)]
+fn time<F>(mut f: F)
+where
+    F: FnMut(),
+{
+    let before = Instant::now();
+    for _ in 0..REPS {
+        f();
```

Comment:
Should we black box the call to `f`, just in case the compiler is clever?

Reply:
Are we sure?

For example, in `bench_normal_no_read` if the compiler can get rid of the asserts, where's the side effect?

I don't think it would hurt to add a black box to be totally sure.

Reply:
It can't get rid of them because the virtual function call means that the compiler can't reason about what the `m` function will do. [If it got rid of these calls, all the benchmark times would be the same!]

Reply:
I understand what you are saying, but would it hurt to add a black box just to be *really* sure?

Reply:
If we take it to that extreme, everything needs to be black boxed. We have to use our discretion.

Reply:
OK, well I would have used a black box, but I won't push any harder.